### PR TITLE
WIP: Support for autogenerated mcu cores

### PR DESCRIPTION
**Pull request by dylanmckay**
It should be possible to do something like this

```rust
extern crate arduino;
use arduino::{Pin, HardwareSpi};
use arduino::cores::atmega328p;

trait Netlist {
    type PushButton: Pin;
    type Led: Pin;
    type Spi: HardwareSpi;
}

struct ATmega328p;

impl Netlist for ATmega328p {
    type PushButton = atmega328p::PORTB3;
    type Led = atmega328p::PORTB4;
    type Spi = atmega328p::SPI;
}

fn start<N: Netlist>() {
    N::PushButton::set_input();
    N::Led::set_output();
    N::Spi::setup_master();

   loop {
        // delay
        M::Led::set_high();
        // delay
        M::Led::set_low();
    }
}
```

This would allow users to easily mix-and-match supported microcontrollers, provided the microcontroller has enough GPIOs, etc.

### CODE REVIEW

File: src/spi.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
+use {Register, Pin};
+
+
+/// An SPI module.
+///
+/// Information at
+/// http://maxembedded.com/2013/11/the-spi-of-the-avr/
+pub trait HardwareSpi {
+    /// Master-in slave-out pin.
+    type MISO: Pin;
```
Comment (shepmaster) [inline src/spi.rs]: I don't know if this is a losing battle, but I really was hoping we could avoid having acronym soup. I'm happy that there are comments, but I would rather it be 

```rust
type MasterInSlaveOut: Pin;
```

And now the comment is encoded in the code. We can then add comments describing the why and the how, instead of the *what*.

The reason I think it might be a losing battle is because there's going to be inertia from all sides to "just do what Arduino / C / Atmel does". The reason I don't think that's fully valid is because we want to do things as Rust-like as we can.

File: src/spi.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    #[inline(always)]
+    fn send_receive(byte: u8) -> u8 {
+        Self::SPDR::write(byte);
+        Self::SPSR::wait_until_set(spsr::SPIF);
+        Self::SPDR::read()
+    }
+}
+
+/// Constants for the control register.
+#[allow(dead_code)]
```
Comment (shepmaster) [inline src/spi.rs]: Why is this needed? Should it be public?

File: src/pin.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
+use {DataDirection, Register};
+
+/// An IO pin.
+pub trait Pin {
+    /// The associated data direction registerr.
```
Comment (shepmaster) [inline src/pin.rs]: extra `r`

File: src/reg.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
+use core::{cmp, convert, ops};
+
+pub trait RegVal : Copy + Clone +
```
Comment (shepmaster) [inline src/reg.rs]: Is there are reason `RegVal` needs to be abbreviated?

File: build.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    let current_mcu = avr_mcu::current::mcu()
+        .expect("no target cpu specified");
+    generate_cores(&[current_mcu]).unwrap()
+}
+
+fn generate_cores(mcus: &[Mcu]) -> Result<(), io::Error> {
+    for mcu in mcus {
+        generate_core_module(mcu).expect("failed to generate mcu core");
```
Comment (shepmaster) [inline build.rs]: This returns a result; do we just want to propagate?

Comment (shepmaster) [reply build.rs]: Or perhaps not until we have an error type to keep the extra information from the `expect` text

File: src/spi.rs (hunk 1 above)
Comment (dylanmckay) [reply src/spi.rs]: I think this is a good idea.

File: src/pin.rs (hunk 3 above)
Comment (dylanmckay) [reply src/pin.rs]: Fixed

File: src/spi.rs (hunk 2 above)
Comment (dylanmckay) [reply src/spi.rs]: I don't want to expose everything just yet because the design isn't 100%, will be public eventually though

File: src/reg.rs (hunk 4 above)
Comment (dylanmckay) [reply src/reg.rs]: Fixed

**Comment by pusherofbrooms** [comment]
There's some great stuff in this PR. Is there any help needed to get it across the finish line? I'd like to use some of the features in https://github.com/pusherofbrooms/rust-arduino-examples without using the extra special branch.
I like that this PR eliminates the std boiler plate.

**Comment by dylanmckay** [comment]
> Is there any help needed to get it across the finish line?

At the moment, it adds autogenerated core functionality side-by-side with the current arduino master. It's a bit strange having two different implementations of the library, and so I think that if we can get feature parity with current master with the new implementation, that should be good.

IIRC, I did most of the work to get it feature complete - I think the SPI stuff is still to be done?

**Comment by dylanmckay** [comment]
Alright, this is now feature complete with the current master branch as far as I can tell.

Still need to do some cleanups, and the whole `serial` module is still using the old atmega328-specific constant tables (rather than the autogenerated core stuff). The serial module can stay that way for now though

**Comment by dylanmckay** [comment]
This is ready for some more eyes again @shepmaster @pusherofbrooms @Restioson

**Comment by dylanmckay** [comment]
I'm going to merge this now to stop faffing about. It's much easier for people to build utop what is here when it is in master, rather than my private fork. I've looked over it locally, cleaned it up a lot. It should work on https://docs.rs too, automatically assuming atmega328p.

**Comment by dylanmckay** [comment]
Docs on [docs.rs](https://docs.rs/ruduino/0.1.2/ruduino/).
